src/Microsoft.AspNetCore.Mvc.Core/Internal/PhysicalFileResultExecutor.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.Mvc.Core;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Mvc.Internal
+{
+    public class PhysicalFileResultExecutor : FileResultExecutor
+    {
+        private const int DefaultBufferSize = 0x1000;
+        private readonly ILogger _logger;
+
+        public PhysicalFileResultExecutor(ILoggerFactory loggerFactory)
+        {
+            if (loggerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _logger = loggerFactory.CreateLogger<PhysicalFileResultExecutor>();
+        }
+
+        public Task ExecuteAsync(ActionContext context, PhysicalFileResult result)
+        {
+            SetHeaders(context, result);
+            _logger.FileResultExecuting(result.FileDownloadName);
+
+            return WriteFileAsync(context, result);
+        }
+
+        private async Task WriteFileAsync(ActionContext context, PhysicalFileResult result)
+        {
+            var response = context.HttpContext.Response;
+
+            if (!Path.IsPathRooted(result.FileName))
+            {
+                throw new NotSupportedException(Resources.FormatFileResult_PathNotRooted(result.FileName));
+            }
+
+            var sendFile = response.HttpContext.Features.Get<IHttpSendFileFeature>();
+            if (sendFile != null)
+            {
+                await sendFile.SendFileAsync(
+                    result.FileName,
+                    offset: 0,
+                    count: null,
+                    cancellation: default(CancellationToken));
+            }
+            else
+            {
+                var fileStream = GetFileStream(result.FileName);
+
+                using (fileStream)
+                {
+                    await fileStream.CopyToAsync(response.Body, DefaultBufferSize);
+                }
+            }
+        }
+
+        protected virtual Stream GetFileStream(string path)
```

Comment:
Does this have to be `protected virtual`? It's not the most convenient of extensibility points. If it is, we should document the type + method.